Unit test for BitUtil Class and Issue #221 Addition of Convenience methods for get/pur char datatypes

David-Wales opened the pull request:
Adding a unit test for a small utility in the Java version of the project And issue #221.  First commit so any feedback would be greatly welcomed!

mjpt777 commented on examples/java/uk/co/real_logic/sbe/examples/ExampleUsingGeneratedStub.java:
```diff
@@ -119,7 +119,7 @@ public static int encode(final Car car, final UnsafeBuffer directBuffer, final i
            .modelYear(2013)
            .available(BooleanType.TRUE)
            .code(Model.A)
-           .putVehicleCode(VEHICLE_CODE, srcOffset);
```
The example was to show the difference between using Strings and raw byte arrays. The Strings are more usable but introduce garbage. Many in HFT want their applications to be garbage free to avoid the subsequent pauses.

David-Wales replied:
That makes sense. I can remove this from the example. Does the generator code for these convenience method look ok?

Thank you!

mjpt777 commented on test/java/uk/co/real_logic/sbe/codec/java/BitUtilTest.java:
```diff
@@ -0,0 +1,84 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+public class BitUtilTest 
```
Why reset this way rather than call the wrap methods?

David-Wales replied:
I'm not sure.  I wrote a small test for it as it was the smallest component I could initially find. But on closer inspection I'm not sure this class (or this method) is even used anywhere.

I could remove the class and test if that would be preferable?

Cheers!

mjpt777 replied:
Removing this makes sense.

David-Wales commented:
So I've removed the unused BitUtil class and tests I added for it. I've also reverted the main examples back to using byte by byte access to char data types. I have left the extra code on the Java Generator  that generates these convenience methods though, as per the issue.

RichardWarburton commented:
I think theres a bitutil in agrona which should be used. So maybe its worth
looking at that?

mjpt777 commented:
I think there is a major flaw with treating the char arrays as strings in SBE. The spec says a `characterEncoding` attribute can be set, it defaults to "UTF-8". Since the array is fixed size and UTF-8 is a variable byte char encoding then this does not match. It feels like char arrays should be restricted to ASCII to ensure a one-for-one byte to char mapping.

Maybe @donmendelson has a view on this?

donmendelson commented:
Some issues with character arrays and strings were clarified in the RC3 spec (I expect it soon to be named final draft of SBE version 1.0.)

Section 2.7.2.1	"Encoding specifications for fixed-length character array"
````
US-ASCII is the default encoding of character arrays to conform to usual FIX values. The characterEncoding attribute may be specified to override encoding. 
````
